### PR TITLE
E2EE Secure ECDH key exchange for DMs

### PR DESCRIPTION
**Pull request by Mega-Mewthree**
I added a few functions to E2EE.js to create secure ECDH key exchanges and obtain a shared password for use with AES by swapping public keys. The key exchanges don't get saved to disk, so the exchange must occur without either person restarting their client after generating their keys. This should make it harder for the original keys to be recovered.

- `createKeyExchange(userID)`: Create a key exchange with another user. Accepts the user ID of the user you are swapping public keys with. Returns the public key that you should give to the other user.

- `publicKeyFor(userID)`: Returns the public key that you should send to the other user. Used when you need to access the public key you created in case you forgot to copy it or something.

- `computeSecret(userID, otherKey)`: Accepts the user ID of the user you swapped keys with, and the other user's public key. Returns a password that can be used for AES. Both users will obtain the same password from this function. Also deletes the key exchange.

### CODE REVIEW

**Comment by Jiiks** [comment]
So many conflicts now

**Comment by Mega-Mewthree** [comment]
Merge conflicts should be fixed now.

File: client/src/builtin/E2EE.js [hunk 1]
```diff
@@ -78,6 +80,16 @@ export default new class E2EE extends BuiltinModule {
         return haveKey.value.value;
     }
 
+    setKey(channelId, key) {
+        const items = Settings.getSetting('security', 'e2eedb', 'e2ekvps').items;
```
Comment (Jiiks) [inline client/src/builtin/E2EE.js]: this.setting

File: client/src/builtin/E2EE.js [hunk 2]
```diff
@@ -78,6 +80,16 @@ export default new class E2EE extends BuiltinModule {
...
+        const index = items.findIndex(kvp => kvp.value.key === channelId);
+        if (index > -1) {
+          items[index].value = {key: channelId, value: key};
+          return;
+        }
+        Settings.getSetting('security', 'e2eedb', 'e2ekvps').addItem({ value: { key: channelId, value: key } });
```
Comment (Jiiks) [inline client/src/builtin/E2EE.js]: this.setting

File: client/src/builtin/E2EE.js [hunk 3]
```diff
@@ -264,6 +276,32 @@ export default new class E2EE extends BuiltinModule {
         MonkeyPatch('BD:E2EE', cta.component.prototype).before('handleSubmit', this.handleChannelTextAreaSubmit.bind(this));
     }
 
+    get ecdh() {
+        if (!this._ecdh) this._ecdh = {};
+        return this._ecdh;
+    }
+
```
Comment (Jiiks) [inline client/src/builtin/E2EE.js]: `return this._ecdh || (this._ecdh = {})`

File: client/src/builtin/E2EE.js [hunk 4]
```diff
@@ -264,6 +276,32 @@ export default new class E2EE extends BuiltinModule {
...
+    createKeyExchange(dmChannelID) {
+        this.ecdh[dmChannelID] = crypto.createECDH('secp521r1');
+        return this.ecdh[dmChannelID].generateKeys('base64');
+    }
+
```
Comment (Jiiks) [inline client/src/builtin/E2EE.js]: Move crypto stuff to Security module

File: client/src/builtin/E2EE.js [hunk 5]
```diff
@@ -264,6 +276,32 @@ export default new class E2EE extends BuiltinModule {
...
+    publicKeyFor(dmChannelID) {
+        return this.ecdh[dmChannelID].getPublicKey('base64');
+    }
+
+    computeSecret(dmChannelID, otherKey) {
```
Comment (Jiiks) [inline client/src/builtin/E2EE.js]: Move crypto stuff to Security module

File: client/src/builtin/E2EEComponent.vue [hunk 6]
```diff
@@ -82,6 +88,30 @@
                     return;
                 }
                 Toasts.success('New messages will be encrypted');
+            },
+            generatePublicKey() {
+                const dmChannelID = location.pathname.split("/")[3];
```
Comment (Jiiks) [inline client/src/builtin/E2EEComponent.vue]: Use DiscordApi for getting channelid `DiscordApi.currentChannel.id`

File: client/src/builtin/E2EEComponent.vue [hunk 7]
```diff
@@ -82,6 +88,30 @@
...
+                const publicKeyMessage = `My public key is: \`${E2EE.createKeyExchange(dmChannelID)}\`. Please give me your public key if you haven't done so and add my public key by pasting it in the chat textbox, right clicking the lock icon, and selecting \`Receive Public Key\`.`;
+                const chatInput = document.getElementsByClassName('da-textArea')[0];
+                chatInput.value = publicKeyMessage;
+                const evt = { currentTarget: chatInput };
+                chatInput[Object.keys(chatInput).find(k => k.startsWith('__reactEventHandlers'))].onChange.call(chatInput, evt);
+            },
+            computeSharedSecret() {
+                try {
+                  const dmChannelID = location.pathname.split("/")[3];
```
Comment (Jiiks) [inline client/src/builtin/E2EEComponent.vue]: Use DiscordApi for getting channelid `DiscordApi.currentChannel.id`

File: client/webpack.config.js [hunk 8]
```diff
@@ -37,6 +37,7 @@ module.exports = {
         util: 'require("util")',
         process: 'require("process")',
         net: 'require("net")',
+        crypto: 'require("crypto")',
```
Comment (Jiiks) [inline client/webpack.config.js]: I added it as `node-crypto` after your addition so use that.

File: client/webpack.production.config.js [hunk 9]
```diff
@@ -38,8 +38,10 @@ module.exports = {
         util: 'require("util")',
         process: 'require("process")',
         net: 'require("net")',
+        crypto: 'require("crypto")',
```
Comment (Jiiks) [inline client/webpack.production.config.js]: I added it as `node-crypto` after your addition so use that.

File: client/src/builtin/E2EEComponent.vue [hunk 10]
```diff
@@ -9,7 +9,7 @@
 */
 
 <template>
-    <div class="bd-e2eeTaContainer">
+    <div class="bd-e2eeTaContainer" @contextmenu.prevent="channelType === 'DM' && $refs.ee2eLockContextMenu.open()">
```
Comment (Jiiks) [inline client/src/builtin/E2EEComponent.vue]: This could just be added to the popover

File: client/src/builtin/E2EEComponent.vue [hunk 11]
```diff
@@ -31,6 +31,10 @@
             </template>
         </v-popover>
         <div class="bd-taDivider"></div>
+        <context-menu id="bd-e2eeLockContextMenu" class="bd-e2eeLockContextMenu" ref="ee2eLockContextMenu" v-if="channelType === 'DM'">
```
Comment (Jiiks) [inline client/src/builtin/E2EEComponent.vue]: This could just be added to the popover

**Comment by Mega-Mewthree** [comment]
I moved the context menu stuff to the popover (using random icons for them for now), moved ECDH functions to Security, and fixed the broken stuff.

**Pull request merged**